google_reauth/reauth_creds.py

```diff
@@ -0,0 +1,190 @@
+from oauth2client import transport
+from google_reauth import reauth
+from google_reauth import reauth_errors
+
+from six.moves import http_client
+from six.moves import urllib
+
+REAUTH_NEEDED_ERROR = 'invalid_grant'
+REAUTH_NEEDED_ERROR_INVALID_RAPT = 'invalid_rapt'
+REAUTH_NEEDED_ERROR_RAPT_REQUIRED = 'rapt_required'
+
+logger = logging.getLogger(__name__)
+
+
+class Oauth2WithReauthCredentials(client.OAuth2Credentials):
+    """Credentials object that extends OAuth2Credentials with reauth support.
+
+    This class provides the same functionality as OAuth2Credentials, but adds
+    the support for reauthentication and rapt tokens. These credentials should
+    behave the same as OAuth2Credentials when the credentials don't use rauth.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Create an instance of Oauth2WithReauthCredentials.
+
+        A Oauth2WithReauthCredentials has an extra rapt_token."""
+
+        if 'rapt_token' in kwargs:
+            self.rapt_token = kwargs['rapt_token']
+            del kwargs['rapt_token']
+        super(Oauth2WithReauthCredentials, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def from_json(cls, json_data):
+        """Overrides."""
+
+        data = json.loads(_helpers._from_bytes(json_data))
+        if ((data.get('token_expiry')
+             and not isinstance(data['token_expiry'], datetime.datetime))):
+            try:
+                data['token_expiry'] = datetime.datetime.strptime(
+                    data['token_expiry'], client.EXPIRY_FORMAT)
+            except ValueError:
+                data['token_expiry'] = None
+
+        kwargs = {}
+        for param in ('revoke_uri', 'id_token', 'id_token_jwt',
+                      'token_response', 'scopes', 'token_info_uri',
+                      'rapt_token'):
+            value = data.get(param, None)
+            if value is not None:
+                kwargs[param] = value
+
+        retval = cls(
+          data['access_token'],
+          data['client_id'],
+          data['client_secret'],
+          data['refresh_token'],
+          data['token_expiry'],
+          data['token_uri'],
+          data['user_agent'],
+          **kwargs
+        )
+        retval.invalid = data['invalid']
+        return retval
+
+    @classmethod
+    def from_OAuth2Credentials(cls, original):
+        """Instantiate a Oauth2WithReauthCredentials from OAuth2Credentials."""
+        json = original.to_json()
+        return cls.from_json(json)
+
+    def _generate_refresh_request_body(self):
+        """Overrides."""
+        parameters = {
+            'grant_type': 'refresh_token',
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'refresh_token': self.refresh_token,
+            'rapt': self.rapt_token,
+        }
+        body = urllib.parse.urlencode(parameters)
+        return body
+
+    def _handle_refresh_error(self, http, rapt_refreshed, resp, content):
```

Comment:
This class does a lot of "low-level" stuff.

I would encourage you to take a look at how we implemented credentials in `google-auth`, there's a complete separation.

Low-level code (functional, stateless): https://github.com/GoogleCloudPlatform/google-auth-library-python/blob/master/google/oauth2/_client.py
High-level class (tracks state and calls low-level code): https://github.com/GoogleCloudPlatform/google-auth-library-python/blob/master/google/oauth2/credentials.py

If you do it this way, it'll be much easier to add reauth to google-auth using the same library.

Reply:
Maybe we can postpone the refactoring till we decide to support reauth in google-auth? I'd be happy to do this later.

Reply:
I just don't want to commit to an interface here that we're just going to have to change later. I don't think it's too much to ask for - refactoring now will make this easier to test, easier to reason about, easier to review, and easier to understand when it's hot-potatoed to the next engineer.

Reply:
PTAL